x11rb-async/src/rust_connection/mod.rs

```diff
@@ -0,0 +1,488 @@
+    ) -> Fut<'future, Cookie<'this, Self, R>, ConnectionError>
+    where
+        'this: 'future,
+        'bufs: 'future,
+        'sl: 'future,
+        R: x11rb::x11_utils::TryParse,
+    {
+        Box::pin(async move {
+            // Gain exclusive access to the write end of the stream.
+            let _guard = self.write_lock.lock().await;
+
+            todo!()
+        })
+    }
+
+    fn send_request_with_reply_with_fds<'this, 'bufs, 'sl, 'future, R>(
+        &'this self,
+        _bufs: &'bufs [io::IoSlice<'sl>],
+        _fds: Vec<RawFdContainer>,
+    ) -> Fut<'future, CookieWithFds<'this, Self, R>, ConnectionError>
+    where
+        'this: 'future,
+        'bufs: 'future,
+        'sl: 'future,
+        R: x11rb::x11_utils::TryParseFd,
+    {
+        todo!()
+    }
+
+    fn send_request_without_reply<'this, 'bufs, 'sl, 'future>(
+        &'this self,
+        _bufs: &'bufs [io::IoSlice<'sl>],
+        _fds: Vec<RawFdContainer>,
+    ) -> Fut<'future, VoidCookie<'this, Self>, ConnectionError>
+    where
+        'this: 'future,
+        'bufs: 'future,
+        'sl: 'future,
+    {
+        todo!()
+    }
+
+    fn check_for_raw_error(
+        &self,
+        sequence: SequenceNumber,
+    ) -> Fut<'_, Option<Self::Buf>, ConnectionError> {
+        Box::pin(async move {
+            // Gain exclusive access to the read end of the stream.
+            let _guard = self.read_lock.lock().await;
+
+            self.do_io(PollMode::Readable, |conn| {
+                conn.inner.check_for_raw_error(sequence)
+            })
+            .await
+        })
+    }
+
+    fn discard_reply(
+        &self,
+        sequence: SequenceNumber,
+        kind: x11rb::connection::RequestKind,
+        mode: x11rb_protocol::DiscardMode,
+    ) {
+        self.inner.discard_reply(sequence, kind, mode)
+    }
+
+    fn parse_error(&self, error: &[u8]) -> Result<x11rb::x11_utils::X11Error, ParseError> {
+        self.inner.parse_error(error)
+    }
+
+    fn parse_event(&self, event: &[u8]) -> Result<x11rb::protocol::Event, ParseError> {
+        self.inner.parse_event(event)
+    }
+
+    fn prefetch_extension_information(&self, name: &'static str) -> Fut<'_, (), ConnectionError> {
+        Box::pin(async move {
+            // Gain exclusive access to the write end of the stream.
+            let _guard = self.write_lock.lock().await;
+
+            self.do_io(PollMode::Writable, move |conn| {
+                conn.inner.prefetch_extension_information(name)
+            })
+            .await
+        })
+    }
+
+    fn extension_information(
+        &self,
+        name: &'static str,
+    ) -> Fut<'_, Option<x11rb::x11_utils::ExtensionInformation>, ConnectionError> {
+        Box::pin(async move {
+            // Gain exclusive access to the write end of the stream.
+            let _guard = self.write_lock.lock().await;
+
+            self.do_io(PollMode::ReadAndWritable, |conn| {
+                conn.inner.extension_information(name)
```

Comment:
Okay, now I feel like I misunderstood the "two mutexes" approach. Why doesn't this need the read lock?

I thought that this had something to do with https://docs.rs/async-io/latest/async_io/struct.Async.html#concurrent-io, but now... I am no longer sure and I am more confused.